### PR TITLE
[CU-frwyvw] Set flow instance size

### PR DESCRIPTION
**Pull request by bhperry**
- [x] passes `make lint`
- [x] adds tests to `tests/` (if appropriate)

## What does this PR change?

Adds instance size argument for flow node

## How does this PR improve `prefect-saturn`?

Enables users to select a larger node size for their flow to run on. This means we are not forcing users to farm all of the work out to dask-clusters, they can instead run flows with a local executor and not be constrained by a medium tier node.

### CODE REVIEW

**Comment by bhperry** [comment]
@jameslamb Not sure whether or not to consider this a breaking change. It is an optional arg so all existing code will continue to function as before, however passing the arg on a version of saturn that doesn't support it will break.

**Comment by hhuuggoo** [comment]
Task linked: [CU-frwyvw Allow for different flow node sizes for Prefect Cloud](https://app.clickup.com/t/frwyvw)

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for doing this! I agree with considering this a non-breaking change, and updating to the patch version.

I have a few requests:

1. If it's not too much effort, could you add something like `dask_saturn.describe_sizes()` that people can use to figure out what the valid values of `instance_size` are? I think having this information come from Atlas is better than hard-coding it in the docs. If you think this is too much of a lift or just want to split up the work, could you document an "add a `prefect_saturn.describe_sizes()`" ticket and share it with me? As I'm sure you're used to from working with terraform or CloudFormation, it's not a fun experience when some docs say "oh this argument takes some string from a list of very this-application-specific strings, but we won't tell you where to find that list".
1. Can you please add details in the docstring on `.register_flow_with_saturn()` that explains how to figure out the valid values?
1. Can you add a unit test that covers that we get the expected behavior when you don't pass `isntance_size` into `.register_flow_with_saturn()`? I think the cleanest way to be sure this is working is to leave the current unit tests untouched and to just add a new one that explicitly passes `instance_size`.

**Comment by bhperry** [comment]
1 and 2 sound doable. dask-saturn does something similar, should be able to reuse some of that.

The reason I didn't do much in the way of testing on this is because I don't think the tests will really tell us anything. The request is just being sent to a mock client, which doesn't care what parameters it receives. This would be better to test in integration tests than in unit tests.

**Comment by jameslamb** [comment]
> The reason I didn't do much in the way of testing on this is because I don't think the tests will really tell us anything. The request is just being sent to a mock client, which doesn't care what parameters it receives. This would be better to test in integration tests than in unit tests.

I believe that making breaking changes to a library should always break the tests, because the tests represent your best guess of how users will call your code. I'd expect a future PR that makes `instance_size` required (a breaking change) should break the tests.

However, the point doesn't matter here since it seems like we already accidentally have a test like that. Here's what happens when I make `instance_size` required and run `make unit-tests`

image

So no new stuff you have to do on testing.

File: prefect_saturn/core.py [hunk 1]
```diff
@@ -104,7 +104,7 @@ def _hash_flow(self, flow: Flow) -> str:
         hasher.update(cloudpickle.dumps(identifying_content))
         return hasher.hexdigest()
 
-    def _set_flow_metadata(self, flow: Flow) -> None:
+    def _set_flow_metadata(self, flow: Flow, instance_size: Optional[str] = None) -> None:
```
Comment (jameslamb) [inline prefect_saturn/core.py]: sorry, one more thing I just thought of. Can you please make `instance_size` required here?

In my experience, when you use this pattern of "public methods take in arguments that they pass through to private methods", it's easy to make mistakes in the upstream code and not correctly pass things through. One way to catch those mistakes is to never allow private methods used in such a pattern to take default values.

```suggestion
    def _set_flow_metadata(self, flow: Flow, instance_size: Union[str, None]) -> None:
```

**Review by jameslamb — CHANGES_REQUESTED** [review]

File: prefect_saturn/core.py (hunk 1 above)
Comment (bhperry) [reply prefect_saturn/core.py]: Good call

File: prefect_saturn/core.py [hunk 2]
```diff
@@ -223,14 +235,21 @@ def register_flow_with_saturn(
         * ``.environment``: a ``KubernetesJobEnvironment`` with a ``DaskExecutor``
             is added. This environment will use the same image as the notebook
             from which this code is run.
+
+        Instance size
+        -------------
+
+        Use the prefect_saturn.describe_sizes function to get the available instance_size options.
```
Comment (jameslamb) [inline prefect_saturn/core.py]:
```suggestion
        Use `prefect_saturn.describe_sizes()` to get the available instance_size options.
```

**Review by jameslamb — APPROVED** [review]
Looks great! Love the way you handled this change, super clean.

I left one small style suggestion for the docs.

**Review by jameslamb — CHANGES_REQUESTED** [review]
sorry, meant to mark that "request changes"

**Review by jameslamb — APPROVED** [review]
:rocket:

**Pull request merged**